usr/opt/OSC200/usr/share/man/man3/putwc.3
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: putwc.3,v $ $Revision: 3.2.5.3 $ (DEC) $Date: 1993/06/21 19:26:44 $
.\"
.\" 
.\" (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
.\" ALL RIGHTS RESERVED 
.\" OSF/1 1.2
.\"
.TH putwc 3
.SH NAME
.PP
\*Lputwc\*O, \*Lputwchar\*O, \*Lfputwc\*O \- Write a character to a stream
.SH LIBRARY
.PP
\*OStandard C Library (\*Llibc.a\*O)
.SH SYNOPSIS
.PP
.sS
\*L#include <wchar.h>
.nL
wint_t putwc(
.nL
	wint_t \*Vc\*L,
.nL
	FILE *\*Vstream\*L);
.nL
wint_t putwchar(
.nL
	wint_t \*Vc \*L);
.nL
wint_t fputwc(
.nL
	wint_t \*Vc\*L;
.nL
	FILE *\*Vstream\*L);
.sE
.SH PARAMETERS
.PP
.VL
.LI "\*Vc\*O"
Specifies the wide character to be written.
.LI "\*Vstream\*O"
Points to the output data.
.LE
.SH DESCRIPTION
.PP
.iX "putwc function"
.iX "putwchar function"
.iX "fputc function"
.iX "IO functions" "standard"
.iX "writing out wide characters"
The \*Lputwc(\|)\*O function converts the \*Lwchar_t\*O specified
by the \*Vc\*O parameter to its equivalent multibyte character and
then writes the multibyte character to the \*Vstream\*O parameter.
.PP
The \*Lputwchar(\|)\*O macro works like the \*Lputwc(\|)\*O function,
except that \*Lputwchar(\|)\*O writes the character
to the standard output.
.PP
The \*Lfputwc(\|)\*O function works the same as \*Lputwc(\|)\*O.
.PP
With the exception of \*Lstderr\*O, output streams are, by
default, buffered if they refer to files, or line buffered if they
refer to terminals. The standard error output stream, \*Lstderr\*O,
is unbuffered by default, but using the \*Lfreopen(\|)\*O function
causes it to become buffered or line buffered. Use the \*Lsetbuf(\|)\*O
function to change the stream's buffering strategy.
.SH RETURN VALUES
.PP
Upon successful completion, these functions each return the value
written. If these functions fail, they return the constant \*LWEOF\*O.
.SH ERRORS
.PP
If any of the following conditions occur, the \*Lputwc(\|)\*O,
\*Lfputwc(\|)\*O, and \*Lputwchar(\|)\*O functions set \*Lerrno\*O to
the corresponding value.
.VL
.LI "[EAGAIN]"
The \*LO_NONBLOCK\*O flag is set for the file descriptor
underlying \*Vstream\*O
and the process would be delayed in the write operation.
.LI "[EBADF]"
The file descriptor underlying \*Vstream\*O
is not a valid file descriptor open for writing.
.LI "[EFBIG]"
An attempt was made to write to a file that exceeds the
process's file size limit or the maximum file size.
.LI "[EINTR]"
The read operation was interrupted by a signal that was caught,
and no data was transferred.
.LI "[EIO]"
The implementation supports job control; the process is a member
of a background process group attempting to write to its
controlling terminal; \*LTOSTOP\*O is set; the process is neither ignoring nor
blocking \*LSIGTTOU\*O; and the process group of the process is
orphaned.
.LI "[ENOSPC]"
There was no free space remaining on the device containing
the file.
.LI "[EPIPE]"
An attempt was made to write to a pipe or FIFO that is not open
for reading by any process.
A \*LSIGPIPE\*O signal will also be sent to the process.
.LI "[EILSEQ]"
The wide character code specified by the \*Vc\*O parameter does not
correspond to a valid character.
.LE
.SH RELATED INFORMATION
.PP
Functions:
\*Lgetc\*O(3),
\*Lgetwc\*O(3),
\*Lprintf\*O(3),
\*Lputc\*O(3),
\*Lputs\*O(3),
\*Lwctomb\*O(3).
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/lib/libc/putwc.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: putwc.c,v $ $Revision: 4.2.5.3 $ (DEC) $Date: 1993/10/05 21:03:08 $";
#endif
/*
 * (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
 * ALL RIGHTS RESERVED 
 */
/*
 * OSF/1 1.2
 */
/*
 * FUNCTIONS: putwc 
 *
 * (C) COPYRIGHT International Business Machines Corp. 1989 
 * All Rights Reserved
 *
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 *
 * putwc.c	1.5  com/lib/c/io,3.1,8943 9/9/89 13:30:54
 */

/* name space pollution clean up */
#ifdef _NAME_SPACE_WEAK_STRONG
#include "pollution.h"
#pragma weak putwc = __putwc
#pragma weak fputwc = __putwc
#endif
#ifdef _THREAD_SAFE

/* Assume streams in this module are already locked.
 */
#define	_STDIO_UNLOCK_CHAR_IO
#endif	/* _THREAD_SAFE */

#include <stdio.h>
#include <stdlib.h>
#include <wchar.h>
#include <ctype.h>
#include <limits.h>
#include "stdiom.h"
#include <errno.h>
#include "ts_supp.h"

#ifdef	_THREAD_SAFE
#include "stdio_lock.h"
#endif	/* _THREAD_SAFE */
#undef putwc
#ifdef _NAME_SPACE_WEAK_STRONG
#define putwc __putwc
#endif


/*
 * FUNCTION:    A subroutine version of the macro putc.  This function is
 *              created to meet ANSI C standards.  The putc function writes
 *              the character specified by c (converted to an unsigned char)
 *              to the output stream pointed to by stream, at the position
 *              indicated by the assoicated file poistion indicator for the
 *              stream, and advances the indicator appropriately.
 *
 * RETURN VALUE DESCRIPTION:	
 *              The putwc function returns the character written.  If a write
 *              error occurs, the error indicator for the stream is set and
 *              putc returns WEOF.
 *
 */
wint_t
putwc(wint_t c, FILE *stream)
{
	int	i, rc, ch;
        char	mbstr[MB_LEN_MAX], *mb;
	TS_FDECLARELOCK(filelock)

	TS_FLOCK(filelock, stream);

        if(MB_CUR_MAX == 1) {
		ch = putc(c, stream);
		TS_FUNLOCK(filelock);
		return (ch == EOF ? WEOF : (wint_t) ch);
	}

        if ((rc = wctomb(mbstr, c)) <= 0) {
                stream->_flag |= _IOERR;
                _Seterrno(EILSEQ);
                TS_FUNLOCK(filelock);
                return (WEOF);
        }

        if (stream->_cnt - rc < 0) {
                mb = mbstr;
                for(i = 0; i < rc; i++) {
                        if (putc(*mb++, stream) == EOF) {
				TS_FUNLOCK(filelock);
                                return (WEOF);
			}
		}
		TS_FUNLOCK(filelock);
                return (c);
        }
	memcpy(stream->_ptr,mbstr,rc);
	stream->_ptr += rc;
	stream->_cnt -= rc;
	_BUFSYNC(stream);
	TS_FUNLOCK(filelock);
	return (c);
}
